File: core/src/main/java/no/elg/infiniteBootleg/console/SystemConsoleReader.java
package no.elg.infiniteBootleg.console;

import com.badlogic.gdx.utils.Disposable;
import java.io.Console;
import java.util.Scanner;
import no.elg.infiniteBootleg.Main;
import org.jetbrains.annotations.NotNull;

/** Read input from {@link System#console()} or {@link System#in} if no console exists. */
public class SystemConsoleReader implements Runnable, Disposable {

  private final ConsoleHandler consoleHandler;
  private boolean running;

  public SystemConsoleReader(@NotNull ConsoleHandler consoleHandler) {
    this.consoleHandler = consoleHandler;

    Thread thread = new Thread(this, "System Console Reader Thread");
    thread.setDaemon(true);
    thread.start();
  }

  @Override
  public void run() {
    if (running) {
      return;
    }
    running = true;
    Console console = System.console();
    try (Scanner scanner =
        console != null ? new Scanner(console.reader()) : new Scanner(System.in)) {
      while (running) {
        String read;
        try {
          read = scanner.nextLine();
          Main.inst().getScheduler().executeSync(() -> consoleHandler.execCommand(read));
        } catch (Exception e) {
          dispose();
        }
      }
    }
  }

  @Override
  public void dispose() {
    running = false;
  }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/console/SystemConsoleReader.kt
package no.elg.infiniteBootleg.console

import com.badlogic.gdx.utils.Disposable
import no.elg.infiniteBootleg.main.Main
import java.io.Console
import java.util.Scanner

/** Read input from [System.console] or [System. in] if no console exists.  */
class SystemConsoleReader(private val consoleHandler: InGameConsoleHandler) : Runnable, Disposable {
  private var running = false
  fun start() {
    val thread = Thread(this, "System Console Reader Thread")
    thread.isDaemon = true
    thread.start()
  }

  override fun run() {
    if (running) {
      return
    }
    running = true
    val scanner = System.console()?.run(Console::reader)?.let(::Scanner) ?: Scanner(System.`in`)
    scanner.use { openScanner ->
      while (running) {
        var read: String?
        try {
          read = openScanner.nextLine()
          Main.inst().scheduler.executeSync { consoleHandler.execCommand(read) }
        } catch (e: Exception) {
          dispose()
        }
      }
    }
  }

  override fun dispose() {
    running = false
  }
}
